nall/gdb/server.cpp

```diff
@@ -0,0 +1,535 @@
+  {
+    auto cmdParts = cmd.split(":");
+    auto cmdName = cmdParts[0];
+    char cmdPrefix = cmdName.size() > 0 ? cmdName[0] : ' ';
+
+    if constexpr(GDB_LOG_MESSAGES) {
+      printf("GDB <: %s\n", cmdBuffer.data());
+    }
+
+    switch(cmdPrefix)
+    {
+      case '!': return "OK"; // informs us that "extended remote-debugging" is used
+
+      case '?': // handshake: why did we halt?
+        haltProgram();
+        haltSignalSent = true;
+        return "T05"; // needs to be faked, otherwise the GDB-client hangs up and eats 100% CPU
+
+      case 'c': // continue
+      case 'C': // continue (with signal, signal itself can be ignored)
+        // normal stop-mode is only allowed to respond once a signal was raised, non-stop must return OK immediately
+        handshakeDone = true; // good indicator that GDB is done, also enables exception sending
+        shouldReply = NON_STOP_MODE;
+        resumeProgram();
+        return "OK";
+
+      case 'D': // client wants to detach (Note: VScode doesn't seem to use this, uses vKill instead)
+        requestDisconnect = true;
+        return "OK";
+      break;
+
+      case 'g': // dump all general registers
+        if(hooks.regReadGeneral) {
+          return hooks.regReadGeneral();
+        } else {
+          return "0000000000000000000000000000000000000000";
+        }
+      break;
+
+      case 'G': // set all general registers
+        if(hooks.regWriteGeneral) {
+          hooks.regWriteGeneral(cmd.slice(1));
+          return "OK";
+        }
+      break;
+
+      case 'H': // set which thread a 'c' command that may follow belongs to (can be ignored in stop-mode)
+        if(cmdName == "Hc0")currentThreadC = 0;
+        if(cmdName == "Hc-1")currentThreadC = -1;
+        return "OK";
+
+      case 'k':  // old version of vKill
+        if(handshakeDone) { // sometimes this gets send during handshake (to reset the program?) -> ignore
+          requestDisconnect = true;
+        }
+        return "OK";
+      break;
+
+      case 'm': // read memory (e.g.: "m80005A00,4")
+        {
+          if(!hooks.read) {
+            return "";
+          }
+
+          auto sepIdxMaybe = cmdName.find(",");
+          u32 sepIdx = sepIdxMaybe ? sepIdxMaybe.get() : 1;
+
+          u64 address = cmdName.slice(1, sepIdx-1).hex();
+          u64 count = cmdName.slice(sepIdx+1, cmdName.size()-sepIdx).hex();
+          return hooks.read(address, count);
+        }
+      break;
+
+      case 'M': // write memory (e.g.: "M801ef90a,4:01000000")
+        {
+          if(!hooks.write) {
+            return "";
+          }
+
+          auto sepIdxMaybe = cmdName.find(",");
+          u32 sepIdx = sepIdxMaybe ? sepIdxMaybe.get() : 1;
+
+          u64 address = cmdName.slice(1, sepIdx-1).hex();
+          u64 unitSize = cmdName.slice(sepIdx+1, 1).hex();
+          u64 value = cmdParts.size() > 1 ? cmdParts[1].hex() : 0;
+
+          hooks.write(address, unitSize, value);
+          return "";
```

Comment:
I think 'M' should return "OK" here, according to spec?

Reply:
Ah that's right, interesting that this worked properly then.
Gonna test if fixing it breaks something else.